### PR TITLE
Fix SM120 BF16 GEMM shared-memory reuse ordering

### PR DESCRIPTION
**Pull request by Sunt-ing**
## Summary

Fixes intermittent silent data corruption in SM120 BF16 GEMM caused by a generic-to-async proxy WAR (Write-After-Read) hazard.

### Problem
Consumer warps read smem tiles via the generic proxy and signal the empty barrier without an async proxy fence. Producer TMA loads (async proxy) can subsequently overwrite the smem stage before generic reads complete across all consumer threads.

### Fix
Add `cutlass::arch::fence_view_async_shared()` before the empty-barrier arrival. Placed outside the lane-0 check so all consumer threads issue the fence for their own smem reads.

Targets `nv_dev` (based on `572557e7ae9ad5331b81a1c250f141fba2c57962`).

## Reproduction and validation

Environment: RTX 5090 (SM120), Driver 590.48.01, PyTorch 2.13.0+cu130.

Added `test_bf16_repeatability` in `tests/test_bf16.py` (`[4096, 1024] × [1024, 896]`, seed 233, 6,000 iterations against initial output).

To run the standalone regression:
```bash
cd tests
python -c "from test_bf16 import test_bf16_repeatability; test_bf16_repeatability()"
```
*Note: The corruption is an intermittent race condition; a single baseline run passing does not guarantee correctness.*

### Repeatability (A/B/A Test)
- **Original:** 2 / 6,000 mismatches
- **Patched:** 0 / 6,000 mismatches
- **Reverted:** 2 / 6,000 mismatches

### Performance Overhead
CUDA graph timing on `[4096, 1024] × [1024, 896]` (median per-call time across 9 batches of 200 calls):

| Version | Median Latency |
| :--- | ---: |
| Original | 50.67 μs |
| Patched | 52.01 μs |
| Reverted | 50.68 μs |

Adds ~2.6% latency overhead on this shape from the fence.

### Supplemental End-to-End Inference Validation
*Internal TP2 inference test (weights/prompts omitted; public verification relies on the standalone GEMM test above).*

- **Setup:** 32,640-token prompt, 89 decode steps -> re-prefill -> 16 decode steps vs. continuous generation. Unified prefill/decode attention was held constant; CUDA graphs were disabled.
- **Original / Reverted:** 6 / 12 runs with divergent logprobs.
- **Patched:** 0 / 12 runs diverged (token IDs and logprobs match strictly).

## Related work
- #389: Addressed the same class of generic-to-async proxy hazard in SM100 MegaMoE combine.
- #447: The consumer release path (reviewed at `1f78b85a1fb974cc69ea63e864d817c603688cab`) does not yet include this fence.

### CODE REVIEW

File: deep_gemm/include/deep_gemm/impls/sm120_bf16_gemm.cuh [hunk 1]
```diff
@@ -338,6 +338,8 @@ sm120_bf16_gemm_impl(cd_dtype_t* gmem_d, const cd_dtype_t* gmem_c,
                     compute_kstep(cur);
                 }
 
+                // Order shared-memory reads before the producer reuses this stage.
+                cutlass::arch::fence_view_async_shared();
                 if (lane_idx == 0)
                     empty_barriers[stage]->arrive();
```
[the remark this answers is not in the document]
Comment (Sunt-ing) [reply deep_gemm/include/deep_gemm/impls/sm120_bf16_gemm.cuh:344]: I evaluated the sibling SM120 release paths on `nv_dev` (`572557e7`) with an RTX 5090 (driver 580.105.08):
- **FP8 / FP4 GEMM** (including tail cases and mixed-precision directions): 0 mismatches across 6,000 iterations per case.
- **HC prenorm**: Passed 2,000 runs with 0 mismatches.
- **Dense / Paged MQA**: Passed 2,000 and 1,000 runs, respectively.
- **BMK / BNK**: Non-determinism stems from FP32 atomic accumulation; adding the fence showed no difference in max error against an FP64 reference.

Agreed that this PR should stay scoped strictly to the verified BF16 WAR hazard to avoid speculative changes without repros. Sibling paths can be tracked and aligned in #447 or separate follow-ups.

The BMK/BNK variation is consistent with FP32 atomic accumulation, though I haven't ruled out other causes. The passing tests also don't rule out ordering hazards in the sibling kernels.

File: tests/test_bf16.py [hunk 2]
```diff
@@ -20,6 +20,25 @@
...
+    a = torch.randn(4096, 1024, device='cuda', dtype=torch.bfloat16,
+                    generator=generator)
+    b = torch.randn(896, 1024, device='cuda', dtype=torch.bfloat16,
+                    generator=generator).t()
+    reference = torch.empty(4096, 896, device='cuda', dtype=torch.bfloat16)
+    output = torch.empty_like(reference)
+    deep_gemm.bf16_gemm_nn(a, b, reference)
+    torch.cuda.synchronize()
+    for iteration in range(6000):
+        deep_gemm.bf16_gemm_nn(a, b, output)
```
[the remark this answers is not in the document]
Comment (Sunt-ing) [reply tests/test_bf16.py:37]: I'd keep checking every call because later calls reuse the output buffer and could overwrite a bad result. torch.equal already waits for the result, so the extra synchronization isn't needed for the comparison itself. I'm keeping it for now to match the before/fix/revert runs. Removing it changes the timing, and I'd want to check that the test still catches the race.

File: tests/test_bf16.py [hunk 3]
```diff
@@ -249,6 +268,7 @@ def test_cublaslt_gemm() -> None:
     print(f' > {deep_gemm.__path__}\n')
 
     if get_arch_major() >= 9:
+        test_bf16_repeatability()
```
[the remark this answers is not in the document]
Comment (Sunt-ing) [reply tests/test_bf16.py:272]: The check is inside test_bf16_repeatability() because the PR documents how to call that function directly. A check only in the script's main block wouldn't cover that invocation.

File: tests/test_bf16.py [hunk 4]
```diff
@@ -20,6 +20,25 @@
...
+                    generator=generator)
+    b = torch.randn(896, 1024, device='cuda', dtype=torch.bfloat16,
+                    generator=generator).t()
+    reference = torch.empty(4096, 896, device='cuda', dtype=torch.bfloat16)
+    output = torch.empty_like(reference)
+    deep_gemm.bf16_gemm_nn(a, b, reference)
+    torch.cuda.synchronize()
+    for iteration in range(6000):
+        deep_gemm.bf16_gemm_nn(a, b, output)
+        torch.cuda.synchronize()
```
[the remark this answers is not in the document]
Comment (Sunt-ing) [reply tests/test_bf16.py:38]: An unfixed build can pass this test. The observed mismatch count isn't enough to give a reliable detection probability. I'll add a docstring calling it a best-effort soak test. The before/fix/revert results support the fix, but one passing run doesn't rule out the race.

File: tests/test_bf16.py [hunk 5]
```diff
@@ -20,6 +20,25 @@
...
 
+def test_bf16_repeatability() -> None:
+    if get_arch_major() != 12:
+        return
+    generator = torch.Generator(device='cuda').manual_seed(233)
+    a = torch.randn(4096, 1024, device='cuda', dtype=torch.bfloat16,
+                    generator=generator)
+    b = torch.randn(896, 1024, device='cuda', dtype=torch.bfloat16,
+                    generator=generator).t()
+    reference = torch.empty(4096, 896, device='cuda', dtype=torch.bfloat16)
```
[the remark this answers is not in the document]
Comment (Sunt-ing) [reply tests/test_bf16.py:32]: torch.equal already waits for its boolean result, so the explicit synchronize is redundant for the comparison. I've kept it to match the runs used to validate the fix. I'd want to recheck the test's ability to catch the race before changing that timing. This file also runs as a standalone script, so I'd keep the early return without adding pytest.skip for this test.

**Comment by Sunt-ing** [comment]
@LyricZhao @zheanxu Could you please take a look at this fix? We found this while investigating reproducibility in LLM evaluation with SGLang on RTX 5090s. With a 32K prompt, continuing decode and re-prefilling the same token prefix sometimes produced different logprobs. Tracing the difference through the model led to a BF16 projection where repeated GEMM calls with identical inputs occasionally returned different values. We then reproduced it with the public random-matrix test in this PR.

**Comment by lucifer1004** [comment]
Have you checked if this is covered by #447?

**Comment by Sunt-ing** [comment]
I checked #447, but I couldn’t find the corresponding fix in the SM120 BF16 kernel. At its current head (1f78b85a), the consumer still calls empty_barriers[stage]->arrive() without a preceding async proxy fence.

**Comment by lucifer1004** [comment]
@Sunt-ing Thanks for the careful A/B/A reproduction — confirmed the hazard class. Your fix has now been ported into #447 at `381e2d7`, via the canonical SM120 device-layer repo ([lucifer1004/DeepGEMM-sm120](https://github.com/lucifer1004/DeepGEMM-sm120) **v0.1.5**, `b31a688`), which both SM120 forks vendor byte-for-byte; the same fix is also in the vllm fork's SM120 layer (vllm-project/DeepGEMM#10).

While merging it we audited all SM120 kernels for the same pattern — TMA-produced smem consumed via `ldmatrix`/`ld_shared` (generic proxy), with the empty-barrier arrive handing the stage back to the async proxy — and added the missing `fence_view_async_shared()` at 14 sites across 8 headers: bf16_gemm (your report), bmk_bnk_mn, tf32_hc_prenorm_gemm, fp8_fp4_gemm_1d1d (×3), fp8/fp4 dense MQA logits (empty_kv + empty_q each), and fp8/fp4 paged MQA logits (empty_q stage release + empty_kv each). The sparse MQA kernel is exempt (its producer is legacy cp.async, generic proxy). Upstream's SM100 kernels carry the same fence for the analogous UMMA case ("Release KV smem only after UMMA commits TMEM"), which corroborates the pattern.

For transparency on evidence: on our RTX PRO 6000 a 30k-iteration soak of the affected shapes did not reproduce the corruption either way, so the fix rests on the memory model plus your 5090 reproduction (2/6000 → 0/6000). Perf A/B (bf16 GEMM, fp8 dense/paged MQA, 3 runs each) shows all deltas within run-to-run noise — consistent with your +2.6% report.

**Comment by Sunt-ing** [comment]
Thanks for porting the fix and checking the other kernels!
